src/test/java/com/microsoft/sqlserver/testframework/sqlType/SqlDateTimeOffset.java

```diff
@@ -25,74 +25,137 @@
+    public static boolean returnMinMax = (0 == r.nextInt(5)); // 20% chance of return Min/Max value
+    private static String numberCharSet2 = "123456789";
+
     // TODO: datetiemoffset can extend SqlDateTime2
     // timezone is not supported in Timestamp so its useless to initialize
     // min/max with offset
-    static SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMdd'T'HH:mm:ss.SSSSSSSZ");
-    static String basePattern = "yyyy-MM-dd HH:mm:ss";
-    static ZoneOffset min = ZoneOffset.of("-1400");
-    static ZoneOffset max = ZoneOffset.of("+1400");
-
-    static DateTimeFormatter formatter;
-
     public SqlDateTimeOffset() {
-        super("datetimeoffset",
-                JDBCType.TIMESTAMP /* microsoft.sql.Types.DATETIMEOFFSET */, null, null);
-        try {
-            minvalue = new Timestamp(
-                    dateFormat.parse((String) SqlTypeValue.DATETIMEOFFSET.minValue).getTime());
-            maxvalue = new Timestamp(
-                    dateFormat.parse((String) SqlTypeValue.DATETIMEOFFSET.maxValue).getTime());
-        }
-        catch (ParseException ex) {
-            fail(ex.getMessage());
-        }
+        super("datetimeoffset", JDBCType.TIMESTAMP /* microsoft.sql.Types.DATETIMEOFFSET */, null, null);
+        minvalue = Timestamp.valueOf((String) SqlTypeValue.DATETIMEOFFSET.minValue);
+        maxvalue = Timestamp.valueOf((String) SqlTypeValue.DATETIMEOFFSET.maxValue);
         this.precision = 7;
         this.variableLengthType = VariableLengthType.Precision;
         generatePrecision();
-        formatter = new DateTimeFormatterBuilder().appendPattern(basePattern)
-                .appendFraction(ChronoField.NANO_OF_SECOND, 0, this.precision, true)
-                .appendOffset("+HH:mm", "Z").toFormatter();
+
     }
 
     public Object createdata() {
-        Timestamp temp = new Timestamp(ThreadLocalRandom.current()
-                .nextLong(((Timestamp) minvalue).getTime(), ((Timestamp) maxvalue).getTime()));
-        temp.setNanos(0);
-        String timeNano = temp.toString().substring(0, temp.toString().length() - 1)
-                + RandomStringUtils.randomNumeric(this.precision);
-
-        // generate random offset values
-        int offsetSeconds = ThreadLocalRandom.current().nextInt(min.getTotalSeconds(),
-                max.getTotalSeconds() + 1);
-
-        // trim the seconds from +HH:mm:ss
-        if (0 != offsetSeconds % 60) {
-            timeNano = timeNano
-                    + ZoneOffset.ofTotalSeconds(offsetSeconds).toString().substring(0, 6);
+        return generateDatetimeoffset(this.precision);
+    }
+
+    public Object generateDatetimeoffset(Integer precision) {
+        if (null == precision) {
+            precision = 7;
+        }
+
+        DateTimeOffset maxDTS = calculateDateTimeOffsetMinMax("max", precision, "9999-12-31 23:59:59");
+        DateTimeOffset minDTS = calculateDateTimeOffsetMinMax("min", precision, "0001-01-01 00:00:00");
+
```

Comment:
It will be better to move these calls into SqlDateTimeOffset() and store them in minvalue and maxvalue. So we don't have to calculate the same value over and over while generating random datetimeoffset value